In [4]:
import tensorflow as tf

w=tf.Variable([.2])  # initial weight, it can be anything
b=tf.Variable([-.2]) # initial bias, it can be anything

x=tf.placeholder(tf.float32)  # input 
y=tf.placeholder(tf.float32)  # y is for holding desired output

linear_model=w*x + b    #calculating actual y


In [5]:
init = tf.global_variables_initializer()  # initalizing variables


s=tf.Session() # setting up session for computation

s.run(init) 
print(s.run(w)) 
print(s.run(b))


[0.2]
[-0.2]


In [6]:
print("actual output is")
print(s.run(linear_model,{x:[1,2,3,4], y:[0,-1,-2,-3]})) # applying this formula w*x + b with initial value of weight and bias, y is the desired output


actual output is
[0.         0.2        0.40000004 0.6       ]


# Lets calculate the error after 1st epoch

In [8]:
squared_deltas=tf.square(linear_model -y)  # used to square the error
loss = tf.reduce_sum(squared_deltas) # reduce_sum is used to sum all the squared errors and give us rmse


print("loss or Error is ")
print(s.run(loss, {x:[1,2,3,4], y:[0,-1,-2,-3]}))   #given input and output values here because we have used placeholder


loss or Error is 
20.16


# error is 20.16 . So lets update the weights and check error again

In [43]:
# now we want to reduce the rmse so we will use gradient descent optimizer

optimizer = tf.train.GradientDescentOptimizer(0.01) # 0.01 is learning rate
train = optimizer.minimize(loss)
s.run(init)
for i in range(1000):  # 1000 is training epoch
 s.run(train, {x:[1,2,3,4,],y:[0,-1,-2,-3]})


In [37]:
print(s.run([w,b])) # we will get values for weight and bias


[array([-0.99999714], dtype=float32), array([0.99999166], dtype=float32)]


In [39]:
s.run(linear_model, {x:[1,2,3,4], y:[0,-1,-2,-3]})


array([-5.4836273e-06, -1.0000026e+00, -1.9999998e+00, -2.9999969e+00],
      dtype=float32)

these values are near to the desired output

In [42]:
s.run(loss, {x:[1,2,3,4], y:[0,-1,-2,-3]})



4.6611603e-11

# rmse after gradient descent optimizer is 4.6611603e-11. Earlier we were getting 20.16